### ISSUE TITLE
ParticleLayer breaks in MapView with zoom level > 12, shows long white horizontal lines on Mac M1

### ISSUE THREAD

**Issue by saintsjd** [weatherlayers/deck.gl-particle#5]
DeckGlParticle.ParticleLayer produces a broken particle layer if H3HexagonLayer is added to list of layers.

Steps to reproduce

1. Edit map.html in docs folder and add

```
<script src="https://unpkg.com/h3-js"></script>
```

2. In map.html, replace `layers` list with 

```
            layers: [
              new deck.TileLayer({
                id: "basemap",
                data: "https://cartodb-basemaps-a.global.ssl.fastly.net/dark_all/{z}/{x}/{y}.png",
                minZoom: 0,
                maxZoom: 19,
                tileSize: 256,
                renderSubLayers: (props) => {
                  const {
                    bbox: { west, south, east, north },
                  } = props.tile;
                  return new deck.BitmapLayer(props, {
                    data: null,
                    image: props.data,
                    bounds: [west, south, east, north],
                  });
                },
              }),
              new DeckGlParticle.ParticleLayer({
                id: "particle",
                maxAge: config.particle.maxAge,
                numParticles: config.particle.numParticles,
                image: "wind_data.png",
                speedFactor: config.particle.speedFactor,
                color: config.particle.color,
                width: config.particle.width,
                opacity: config.particle.opacity,
                animate: config.particle.animate,
                getPolygonOffset: ({ layerIndex }) => [0, -1000], // layerIndex is per view
              }),
              new deck.H3HexagonLayer({
                id: "hexagon-layer",
                data: [
                  {
                    hex: "88283082b9fffff",
                  },
                ],
                // pickable: true,
                // wireframe: false,
                // filled: true,
                // stroked: true,
                extruded: false,
                getHexagon: (d) => d.hex,
                getFillColor: (d) => [255, 255, 0],
              }),
            ],

```

4. Keep zooming in until particle layer breaks

Result:
<img width="1021" alt="Screen Shot 2022-03-03 at 4 40 50 PM" src="https://user-images.githubusercontent.com/24510/156671619-a7ff886d-417e-453d-9ca1-c43eea411cec.png">

Expected result:
<img width="1012" alt="Screen Shot 2022-03-03 at 4 42 06 PM" src="https://user-images.githubusercontent.com/24510/156671687-781bd5e2-22c4-4a1a-b2a5-468e654cc391.png">

**Comment by zakjan** [comment]
Interesting, thanks for the report, I'll take a look.

**Comment by saintsjd** [comment]
Adding a geojson layer causes the same issue. So issue is not h3 related. For example:

```
new deck.GeoJsonLayer({
                id: "geojson-layer",
                data: {
                  type: "FeatureCollection",
                  features: [
                    {
                      type: "Feature",
                      properties: {},
                      geometry: {
                        type: "Polygon",
                        coordinates: [
                          [
                            [-58.963623046875, -34.88142481679756],
                            [-58.17260742187501, -34.88142481679756],
                            [-58.17260742187501, -34.384246040152185],
                            [-58.963623046875, -34.384246040152185],
                            [-58.963623046875, -34.88142481679756],
                          ],
                        ],
                      },
                    },
                  ],
                },
                pickable: true,
                stroked: false,
                filled: true,
                extruded: true,
                pointType: "circle",
                lineWidthScale: 20,
                lineWidthMinPixels: 2,
                getFillColor: [160, 160, 180, 200],
                getPointRadius: 100,
                getLineWidth: 1,
                getElevation: 30,
              }),
```

**Comment by zakjan** [comment]
Yes, I saw it too once without any additional layers, but no matter I tried, I couldn't replicate it. Possibly it could be hardware dependant. However, if your system is Mac M1 (guessing from your mention of arm64 in the other ticket), we should see the same result. Could you try it with deck.gl 8.6.0, to check if it was caused by the upgrade to 8.7.0?

**Comment by saintsjd** [comment]
I just tried with 8.6.0 and the issue is not resolved (verified that 8.6.0 is being used in network tab). 

I am on an m1. I do believe this is hardware dependent. We tested on a number of windows machines and we do not see this issue. Another mac laptop (not m1) seems to show the issue even without the added geojson layer. 

I can reproduce on my machine. If I comment out the geojson layer, then the issue goes away. If I add the geojson layer the issue is always there when I zoom past z12. 

I will keep trying to narrow down the cause. So far no luck. Let me know if there is anything else you would like me to try.

**Comment by zakjan** [comment]
I can indeed replicate it now on Mac M1, even without the additional layer. I haven’t found the cause yet, but it’s on the priority list.

**Comment by zakjan** [comment]
Observation: it happens in MapView only.

Zooming in GlobeView works fine until ~15.5, when the particle trails become straight. This is where I suspect we hit the precision issue #3.

**Comment by zakjan** [comment]
It's directly related to auto-offset mode which triggers at zoom 12. https://github.com/visgl/deck.gl/blob/3c995f53a1b2df3f0196dbbe8768bcc349fec8da/modules/core/src/viewports/viewport.ts#L180-L182

Changing to a different zoom threshold causes the bug to appear at the updated zoom.

**Comment by saintsjd** [comment]
I will read more about auto-offest. I am not familiar with it. Good find!

**Comment by zakjan** [comment]
I have fixed this issue in https://github.com/weatherlayers/deck.gl-particle/commit/f40e5a9942f7a794ee94dc9b41772dfbcad04ff5, released as 1.0.6.

The root cause was that in the auto-offset mode, 64-bit low buffers are used, but ParticleLayer didn't provided them. On most of devices they were automatically filled with 0, but on Mac M1 an explicit buffer filled with 0 must be provided.

@saintsjd Could you test 1.0.6 on your devices?

**Comment by zakjan** [comment]
@maxwell-oroark I'm sorry about it, please try out released 1.0.7

**Comment by zakjan** [comment]
Err, one more typo, fixed in 1.0.8

**Comment by zakjan** [comment]
Thanks for your testing! The direction change is caused by precision issues. Yes, disabling the layer in high zoom is the current workaround.
